smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php

```diff
@@ -0,0 +1,277 @@
+
+/**
+ * Smaily module admin configuraiton form.
+ */
+class SmailyForDrupalAdminForm extends ConfigFormBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getEditableConfigNames() {
+    return [
+      'smaily_for_drupal.adminsettings',
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId() {
+    return 'smaily_for_drupal_admin_form';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $config = $this->config('smaily_for_drupal.adminsettings');
+
+    // Wrapper for editing and returning form data with ajax.
+    $form['smaily_container'] = [
+      '#type' => 'container',
+      '#attributes' => [
+        'id' => 'smaily-wrapper',
+      ],
+    ];
+
+    $form['smaily_container']['credentials'] = [
+      '#type' => 'fieldset',
+      '#title' => $this->t('API credentials'),
+    ];
+
+    $form['smaily_container']['credentials']['message'] = [
+      '#type' => 'markup',
+      '#markup' => '<div class="smaily_message"></div>',
+    ];
+
+    $form['smaily_container']['credentials']['smaily_domain'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Domain'),
+      '#default_value' => $config->get('smaily_domain', ''),
+      '#attributes' => [
+        'placeholder' => $this->t('For example "demo" from https://demo.sendsmaily.net/'),
+      ],
+      '#required' => TRUE,
+    ];
+
+    $form['smaily_container']['credentials']['smaily_username'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Username'),
+      '#default_value' => $config->get('smaily_username', ''),
+      '#required' => TRUE,
+    ];
+
+    $form['smaily_container']['credentials']['smaily_password'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Password'),
+      '#default_value' => $config->get('smaily_password', ''),
+      '#required' => TRUE,
+
+    ];
+
+    $form['smaily_container']['credentials']['validate_credentials'] = [
+      '#type' => 'button',
+      '#value' => $this->t('Validate credentials'),
+      '#ajax' => [
+        'event' => 'click',
+        'callback' => '::validateCredentials',
+        'wrapper' => 'smaily-wrapper',
+      ],
+      // Limit errors here, validateCredentials will highlight errors in autoresponder/title.
+      '#limit_validation_errors' => [],
+    ];
+
+    $username = $config->get('smaily_username');
+    $password = $config->get('smaily_password');
+    $domain = $config->get('smaily_domain');
+
+    // Fetch and update autoresponders every time form is built.
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL,
+      'https://' . $domain . '.sendsmaily.net/api/workflows.php?trigger_type=form_submitted'
+    );
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
+    curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+    curl_setopt($ch, CURLOPT_USERPWD, "$username:$password");
+    $autoresponders = json_decode(curl_exec($ch), TRUE);
```

Comment:
What happens if `curl_exec` returns `FALSE`?

Reply:
I tried `json_decode(FALSE, TRUE)` it returned NULL
`if (!empty($autoresponders))` will be skipped

Reply:
Is it bad not to handle a FALSE message and rely on json_decode returning NULL